rfc/004-reverse-sync.md

```diff
@@ -0,0 +1,198 @@
+`A.Height <= max_historical_height` and `A.Time <= max_historical_time`.
+
+Upon successfully reverse syncing, a node can now safely continue. As this
+feature is only used as part of state sync, one can think of this as merely an
+extension to it.
+
+In the future we may want to extend this functionality to allow nodes to fetch
+historical blocks for reasons of accountability or data accessibility.
+
+### Verification
+
+ReverseSync is used to fetch and verify the following data structures:
+- `Header`
+- `ValidatorSet`
+
+Nodes will need to verify these prior blocks. This can be achieved by first
+retrieving the header at the base height from the block store. The node then
+checks that the `LastBlockID` corresponds with the hash calculated from the
+header in the new block directly below:
+
+```go
+header[height].LastBlockID == hash(header[height-1])
+```
+
+The node can therefore trust the new header. It then fetches the validator set
+and validates it as such:
+
+```go
+header[height-1].ValidatorHash == hash(validatorSet[height-1])
+```
+
+### Design
+
+This section will provide a high level overview of some of the more important
+characteristics of the design, saving the more tedious details as an ADR. This
+section is divided into two approaches that can be used to achieve the defined
+backfill mechanism based on the two networking protocols: RPC or P2P.
+
+#### RPC
+
+The aforementioned data is already available via the following RPC endpoints:
+`/blockchain` or `/commit` for `Header`'s' and `/validators` for
+`ValidatorSet`'s'. It may be the best option to fetch and verify these resources
+over RPC. Statesync already requires a list of RPC endpoints, so these could be
+reused.
+
+#### P2P
+
+A more congruent approach to the current infrastructure would be to request and
+send the respective resources over the p2p protocol. This would mean the
+addition of two new messages.
+
+```proto
+message ValidatorSetRequest {
+  uint64 height = 1;
+}
+```
+
+```proto
+message ValidatorSetResponse {
+  uint64 height = 1;
+  ValidatorSet validator_set = 2;
+}
+```
+
+In addition, Tendermint currently sends messages for entire blocks only. This
+would be inconvenient if we only wanted to acquire the header (although there
+may be benefits in saving the entire block). Hence we may want to consider
+increasing the granularity of the request of resources by separating out
+components such as header, commit and data (includes evidence).
+
+The P2P path may also enable P2P networked light clients and a state sync that
+also doesn't need to rely on RPC as the adequate data can now be transmitted
+through P2P.
+
+### Termination
+
+ReverseSync draws a lot of parallels with fast sync. An important consideration
+for fast sync that also extends to ReverseSync is termination. ReverseSync will
+finish it's task when one of the following conditions have been met:
+
+1. It reaches a block `A` where `A.Height <= max_historical_height` and
```

Comment:
What we use for as an upper bound of the unbonding period window, now or bft time of the state synced state? In the former case, we have a moving target for termination.

Reply:
bft time of the state synced state